### PR TITLE
refactor: change tree node processing way

### PR DESCRIPTION
**Pull request by grianbrcom**
Syntax tree will be processed several times.

Signed-off-by: Anton Grigorev <anton.grigorev@broadcom.com>

### CODE REVIEW

File: server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java [hunk 1]
```diff
@@ -156,14 +157,19 @@ public CobolLanguageEngine(
       timingBuilder.getSyntaxTreeTimer().start();
       analyzeEmbeddedCode(syntaxTree, positionMapping, context.getConstants());
       Node rootNode = syntaxTree.get(0);
-      accumulatedErrors.addAll(rootNode.process());
+      int processCalls = 0;
+      do {
+        accumulatedErrors.addAll(rootNode.process());
+        processCalls++;
+        if (processCalls > PROCESS_CALLS_THRESHOLD) throw new RuntimeException("Infinity loop in tree processing");
+      } while (!rootNode.isProcessed());
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java:165]: Please, extract it to a method.

File: server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java [hunk 2]
```diff
@@ -73,11 +74,20 @@ public void setProgramName(String programName) {
     this.programName = programName;
   }
 
-  @Override
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java:76]: `@Overrride` should remain here.

File: server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java [hunk 3]
```diff
@@ -162,4 +172,8 @@ public void addVariableDefinition(VariableNode node) {
         .flatMap(Collection::stream)
         .collect(toList());
   }
+
+  private enum NodeState {
+    NEW, WAIT_FOR_VARIABLES, DONE;
+  }
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java:178]: Shouldn't it either belong to the Node or be a separate class?

File: server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java (hunk 1 above)
Comment (grianbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java:165]: Agree.

File: server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java (hunk 3 above)
Comment (grianbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java:178]: Each node can have its own stages. I don't expect to have one set of states for all nodes. What if some node requires more than three states?

Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java:178]: Makes sense

**Review by temanbrcom — APPROVED** [review]
🌳

**Pull request merged**